File: Modules/Combat.py
from enum import Enum
import random

# Enums
class Tactic(Enum):
    Default = 0
    Custom = 1
    Panic = 2
    Defensive = 3
    Aggressive = 4
    Combo = 5

# Class for Combat System
class CombatSystem:
    def __init__(self):
        pass

    def initiate(self, attacker, defender, tactic):
        if tactic == Tactic.Custom:
            self.attack(attacker, defender)
        elif tactic == Tactic.Panic:
            self.handle_panic(defender)
        elif tactic == Tactic.Defensive:
            self.defensive_attack(attacker, defender)
        elif tactic == Tactic.Aggressive:
            self.aggressive_attack(attacker, defender)
        elif tactic == Tactic.Combo:
            self.combo_attack(attacker, defender)

    def attack(self, attacker, defender):
        print(f"{attacker.name} attacks {defender.name}.")
        base_damage = random.randint(10, 20)  # Random base damage between 10 and 20
        total_damage = base_damage + attacker.damage_bonus  # Add attacker's damage bonus
        defender.take_damage(total_damage)
        print(f"{defender.name} takes {total_damage} damage. Current health: {defender.health}")

        # Additional logic for gaining experience upon successful attack
        attacker.gain_experience(20)  # Example: Gain 20 experience points per successful attack

    # Defensive tactic: Focus on blocking and counter-attacks
    def defensive_attack(self, attacker, defender):
        print(f"{attacker.name} takes a defensive stance.")
        # Add logic here for defensive actions such as blocking or counter-attacks

    # Aggressive tactic: Focus on overwhelming the opponent with relentless attacks
    def aggressive_attack(self, attacker, defender):
        print(f"{attacker.name} goes on the offensive!")
        # Add logic here for aggressive attacks, possibly with increased damage but higher risk

    # Combo tactic: Execute a series of coordinated attacks for increased damage
    def combo_attack(self, attacker, defender):
        print(f"{attacker.name} unleashes a devastating combo!")
        # Add logic here for executing combo attacks, possibly requiring specific conditions to trigger

    def handle_panic(self, combatant):
        if combatant.health < 35:
            print(f"{combatant.name} is panicking!")
            combatant.conscious = False

# Example usage:
# combat_system = CombatSystem()
# tactic = Tactic.Custom
# attacker = ...
# defender = ...
# combat_system.initiate(attacker, defender, tactic)

File: Modules/Combat.cs
using System;

// Enums
public enum Tactic
{
    Default,
    Custom,
    Panic,
    Defensive,
    Aggressive,
    Combo
}

// Class for Combat System
public class CombatSystem
{
    public CombatSystem() { }

    public void Initiate(Character attacker, Character defender, Tactic tactic)
    {
        switch (tactic)
        {
            case Tactic.Custom:
                Attack(attacker, defender);
                break;
            case Tactic.Panic:
                HandlePanic(defender);
                break;
            case Tactic.Defensive:
                DefensiveAttack(attacker, defender);
                break;
            case Tactic.Aggressive:
                AggressiveAttack(attacker, defender);
                break;
            case Tactic.Combo:
                ComboAttack(attacker, defender);
                break;
            default:
                break;
        }
    }

    public void Attack(Character attacker, Character defender)
    {
        Console.WriteLine($"{attacker.Name} attacks {defender.Name}.");
        Random rand = new Random();
        int baseDamage = rand.Next(10, 21); // Random base damage between 10 and 20
        int totalDamage = baseDamage + attacker.DamageBonus; // Add attacker's damage bonus
        defender.TakeDamage(totalDamage);
        Console.WriteLine($"{defender.Name} takes {totalDamage} damage. Current health: {defender.Health}");

        // Additional logic for gaining experience upon successful attack
        attacker.GainExperience(20); // Example: Gain 20 experience points per successful attack
    }

    // Defensive tactic: Focus on blocking and counter-attacks
    public void DefensiveAttack(Character attacker, Character defender)
    {
        Console.WriteLine($"{attacker.Name} takes a defensive stance.");
        // Add logic here for defensive actions such as blocking or counter-attacks
    }

    // Aggressive tactic: Focus on overwhelming the opponent with relentless attacks
    public void AggressiveAttack(Character attacker, Character defender)
    {
        Console.WriteLine($"{attacker.Name} goes on the offensive!");
        // Add logic here for aggressive attacks, possibly with increased damage but higher risk
    }

    // Combo tactic: Execute a series of coordinated attacks for increased damage
    public void ComboAttack(Character attacker, Character defender)
    {
        Console.WriteLine($"{attacker.Name} unleashes a devastating combo!");
        // Add logic here for executing combo attacks, possibly requiring specific conditions to trigger
    }

    public void HandlePanic(Character combatant)
    {
        if (combatant.Health < 35)
        {
            Console.WriteLine($"{combatant.Name} is panicking!");
            combatant.Conscious = false;
        }
    }
}

// Example Character class
public class Character
{
    public string Name { get; set; }
    public int Health { get; set; }
    public int DamageBonus { get; set; }
    public bool Conscious { get; set; }

    public void TakeDamage(int damage)
    {
        Health -= damage;
    }

    public void GainExperience(int experience)
    {
        // Logic for gaining experience
    }
}

// Example usage:
// CombatSystem combatSystem = new CombatSystem();
// Tactic tactic = Tactic.Custom;
// Character attacker = new Character();
// Character defender = new Character();
// combatSystem.Initiate(attacker, defender, tactic);
